mypyc/genops.py

```diff
@@ -1603,158 +1604,138 @@ def for_loop_helper(self, index: Lvalue, expr: Expression,
                         line: int) -> None:
         """Generate IR for a loop.
 
-        "index" is the loop index Lvalue
-        "expr" is the expression to iterate over
-        "body_insts" is a function to generate the body of the loop.
+        Args:
+            index: the loop index Lvalue
+            expr: the expression to iterate over
+            body_insts: a function that generates the body of the loop
+            else_insts: a function that generates the else block instructions
         """
-        body_block, increment_block = BasicBlock(), BasicBlock()
-        # Block for the else clause, if we need it.
+        # Body of the loop
+        body_block = BasicBlock()
+        # Block that steps to the next item
+        step_block = BasicBlock()
+        # Block for the else clause, if we need it
         else_block = BasicBlock()
+        # Block executed after the loop
         exit_block = BasicBlock()
 
         # Determine where we want to exit, if our condition check fails.
         normal_loop_exit = else_block if else_insts is not None else exit_block
 
-        # Only support 1 and 2 arg forms for now
-        if (isinstance(expr, CallExpr)
-                and isinstance(expr.callee, RefExpr)
-                and expr.callee.fullname == 'builtins.range'
-                and len(expr.args) <= 2):
-
-            condition_block = BasicBlock()
-            self.push_loop_stack(increment_block, exit_block)
-
-            # Special case for x in range(...)
-            # TODO: Check argument counts and kinds; check the lvalue
-            if len(expr.args) == 1:
-                start_reg = self.add(LoadInt(0))
-                end_reg = self.accept(expr.args[0])
-            else:
-                start_reg = self.accept(expr.args[0])
-                end_reg = self.accept(expr.args[1])
-
-            end_target = self.maybe_spill(end_reg)
-
-            # Initialize loop index to 0. Assert that the index target is assignable.
-            index_target = self.get_assignment_target(
-                index)  # type: Union[Register, AssignmentTarget]
-            self.assign(index_target, start_reg, line)
-            self.goto(condition_block)
+        for_gen = self.make_for_loop_generator(index, expr, body_block, normal_loop_exit, line)
 
-            # Add loop condition check.
-            self.activate_block(condition_block)
-            comparison = self.binary_op(self.read(index_target, line),
-                                        self.read(end_target, line), '<', line)
-            self.add_bool_branch(comparison, body_block, normal_loop_exit)
-
-            self.activate_block(body_block)
-            body_insts()
+        self.push_loop_stack(step_block, exit_block)
+        if for_gen.has_combined_step_and_condition():
+            self.goto_and_activate(step_block)
+        else:
+            # We need a separate condition block.
+            condition_block = self.goto_new_block()
```

Comment:
Is the `has_combined_step_and_condition()` stuff just to avoid creating some useless basic blocks? If so we can probably skip it, since I added a basic block cleanup optimization pass as a precursor to the uninitialized variables work. Right now it runs right after refcount insertion but it is going to need to run between genops and uninitialized variable insertion, so we can use it to clean up these blocks and avoid some special case code, I think.